### PR TITLE
8277501: Revisit PathFileObject.getCharContent and friends

### PR DESCRIPTION
**Pull request by archiecobbs**
The method `BaseFileManager.makeByteBuffer()` assumes that the result from`InputStream.available()` is always accurate, but this is not guaranteed.

This patch cleans up this logic by piggybacking on a similar cleanup that was recently done for [JDK-8302514](https://bugs.openjdk.org/browse/JDK-8302514).

This change also makes a couple of other changes:

(1) Adjust the sematics of `ByteBuffer.appendStream()` so that it doesn't also close the input, which is abnormal for a method like this.

(2) Add synchronization to the buffer caching in `BaseFileManager.makeByteBuffer()` because, according to the API documentation for `JavaFileManager`:
> An object of this interface is not required to support multi-threaded access, that is, be synchronized. However, it must support concurrent access to different file objects created by this object. 

`BaseFileManager.makeByteBuffer()` is invoked from `PathFileObject` and so the per-file object thread-safety requirement would apply here.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8277501](https://bugs.openjdk.org/browse/JDK-8277501): Revisit PathFileObject.getCharContent and friends

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12755/head:pull/12755` \
`$ git checkout pull/12755`

Update a local copy of the PR: \
`$ git checkout pull/12755` \
`$ git pull https://git.openjdk.org/jdk.git pull/12755/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12755`

View PR using the GUI difftool: \
`$ git pr show -t 12755`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12755.diff">https://git.openjdk.org/jdk/pull/12755.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/ByteBuffer.java [hunk 1]
```diff
@@ -56,8 +56,13 @@ public ByteBuffer() {
      *  of given size.
      */
     public ByteBuffer(int initialSize) {
-        elems = new byte[initialSize];
-        length = 0;
+        this(new byte[initialSize]);
+    }
+
+    /** Create a new byte buffer using the given array for storage.
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/util/ByteBuffer.java:62]: Generally, the style for doc comments is either a single line, or for a newline after the initial `/**`

**Review by jonathan-gibbons — COMMENTED** [review]
Is it worth renaming the internal `ByteBuffer` class to avoid the name clash?

Just asking.

The counter-argument is that we already deal with `javac.util.List` and `java.util.List`

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/ByteBuffer.java (hunk 1 above)
Comment (archiecobbs) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/util/ByteBuffer.java:62]: Yeah... except not in this file. I was erring on the side of being consistent with the rest of the file (lesser of two evils?)

**Comment by vicente-romero-oracle** [comment]
hi guys, would like to close this one, any other comments?

**Comment by vicente-romero-oracle** [comment]
tier1-4 green

**Comment by archiecobbs** [comment]
Hi @vicente-romero-oracle ,

> would like to close this one, any other comments?

If you're asking me, I think renaming `ByteBuffer` is a reasonable idea but should go into a separate PR (guess I'm trying to avoid mission creep). So I'm ready to wrap this one up but am happy to entertain other opinions.

Thanks.

**Comment by jonathan-gibbons** [comment]
> Hi @vicente-romero-oracle ,
> 
> > would like to close this one, any other comments?
> 
> If you're asking me, I think renaming `ByteBuffer` is a reasonable idea but should go into a separate PR (guess I'm trying to avoid mission creep). So I'm ready to wrap this one up but am happy to entertain other opinions.
> 
> Thanks.

Avoiding creeping missions is a good idea.  One way to help with that can be to file JBS issues as a future reminder.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/ByteBuffer.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/util/ByteBuffer.java:62]: OK, as long as you made a deliberate choice, that's cool.

**Review by jonathan-gibbons — APPROVED** [review]

**Comment by vicente-romero-oracle** [comment]
good, closing this one, thanks guys!

**Pull request closed without merging**